### ISSUE TITLE
Frequent apps

### ISSUE THREAD

**Issue by eevleevs** [micheleg/dash-to-dock#576]
I would like an option to display a certain number of apps from the frequents list instead of predefined locked favorites. This would allow for dynamic usage of the dock space.

**Comment by micheleg** [comment]
I never considered it, or if I did, it was a long time ago. Thanks for the suggestion. It would be something worth exploring, but I'm not sure it would actually work.

Could you checkout this branch:

https://github.com/micheleg/dash-to-dock/tree/experimental_frequent_apps

and tell me what do you think. 
I quikly tried to playing with  this feature,  showing the 5 most used apps instead of the favorites.
It's just a dirty 5 minutes hack to play with it for the moment: some things could be weird and I can't promise it won't crash...

**Comment by eevleevs** [comment]
Thanks, looks promising! I got some frequents, and no crash so far :+1: 

The list is not exactly the one on the menu:
![schermata del 2017-08-24 20-49-08](https://user-images.githubusercontent.com/5012744/29688120-1e6ed68a-890e-11e7-9991-6aa7e1cf3182.png)

and some of the programs that were marked as favourites before I installed the branch now are not displayed (see terminal and tweak-tool, but not chromium):
![schermata del 2017-08-24 20-50-56](https://user-images.githubusercontent.com/5012744/29688121-1eca2a12-890e-11e7-8c18-5ecb7e52b680.png)

Anyway let me change my suggestion: frequent and favourites do not necessarily need to be exclusive. Favourites could still be displayed at the top of the frequents list.

**Comment by micheleg** [comment]
I have updated the  branch with a couple of fixings. The issue with the different apps shown could be due to the fact that it's not clear to me when the list should be updated...

I agree that a favorites + frequent makes more sense.

**Comment by micheleg** [comment]
> Favourites could still be displayed at the top of the frequents list.

Added now.

**Comment by eevleevs** [comment]
Favourites on top seems to work well. Now I have only three recents, not sure if it is connected to updating. In my mind, one update at login could already be sufficient. Otherwise you could set a periodic one. Hourly? Daily?

**Comment by eevleevs** [comment]
Also, it looks like it disabled grouping
![image](https://user-images.githubusercontent.com/5012744/29703293-736ee412-8964-11e7-8214-fc394b8f4011.png)

**Comment by micheleg** [comment]
Yes, I limited to favorites plus the first three frequent applications not already in favorites. The issue with hte order is that it is recalculated upon many events, and i don't think it is vey usable to have launchers unexpectively jumping around. One possibiity could be to keep then constant in the session (but then a session can easily last for a month) 

Other things to consider: 
* should the frequent app be after the running ones instead to avoid having the running apps far to the end of the dock?
* should the frequent apps be shown only if not more than X lanuchers are already shown (i.e. the number shown be dynamic)?

Overall, from your use, how much of an help is having the frequent apps?

**Comment by eevleevs** [comment]
- I am perfectly ok with updating after a certain number of events. Being curious, what is the number?
-  I don't think frequents should come after running. They are in concept similar to favourites, something you pin up to have at hand. Also you would and up with a lot of movement, not sure it would play nice.
- I would just make the number of frequents displayed customisable from the options. Probably easier and more flexible. To say, I would fill the dock with frequents.
- I love the concept and I am definitively going to use it if you finalise it. I am using a similar thing for my android launcher and I find it very practical.

**Comment by micheleg** [comment]
>I am perfectly ok with updating after a certain number of events. Being curious, what is the number?

I meant that there are other unreleted events (e.g. adding a favorite) that trigger a full redraw of the dock, therefore updating the frequent s app too. 

I had the option enabled for the last few days and to be honest I didn't notice the difference. Which is good, because it means that if feel quite natural to look for the launcher on the dock and if it not there use  the show-app icon to pick it from the overview. In a sense the feature is a but reduntant because of this, but I occasionally noticed I was picking an app I normally don't use much and I was pleased to find it there.

Let me know if you have any additional comments. At the moment the implementation is faily dirty. I'll need to find a moment to clean it up before considering including this feature.

**Comment by eevleevs** [comment]
I filled the dock with frequents and removed all favourites, works quite nice. The only glitch I noticed is that I ended up with two file managers. Sometimes, another file manager appears when a transfer is ongoing, not sure if related?

![image](https://user-images.githubusercontent.com/5012744/29858583-5a8fc5e4-8d5e-11e7-9d8b-85f9192f9a59.png)

**Comment by micheleg** [comment]
Fairly strange... thanks for the additional feedback.

**Comment by eevleevs** [comment]
I still have it, let me know if I can investigate somehow.

**Comment by eevleevs** [comment]
Another thought: you might want the possibility to exclude a program from recents. For example, evince is likely to end up in the recents for me, as I often use it from the file manager. But I will rarely launch it directly from the dock, where it will just take space.

"Right click -> exclude from recents" would do. But you would also need a black list somewhere in the options for resetting.

**Comment by eevleevs** [comment]
> The only glitch I noticed is that I ended up with two file managers. Sometimes, another file manager appears when a transfer is ongoing, not sure if related?

Definitively related to file transfer. I made a quite long move to USB tonight and the second file manager is back!

**Comment by micheleg** [comment]
I understand, but I'm not sure if I want to pollute the menus with additional entries. Although this would be limited to those few favorites launchers shown. It wil consider think about this.

Thanks for the additional information about the double icon. I will try to reproduce the bug when I will work actively on this feature, at the moment I want to focus on other things.  Let me know if you happen to find other issues of if you have any other comments, thank you.

**Comment by eevleevs** [comment]
> I understand, but I'm not sure if I want to pollute the menus with additional entries. Although this would be limited to those few favorites launchers shown. It wil consider think about this.

Otherwise, a black list on file would be sufficient, at least for some people.

**Comment by riyaniii** [comment]
@micheleg would this be added? or atleast recent apps?
